File: VehicleVision.Tools.ScreenSketch.VSCode/src/themeColors.ts
// Theme dimension / typography constants
export const Theme = {
    FontFamily: "'Segoe UI', 'Meiryo', sans-serif",
    DefaultFontSize: 12,
    TitleBarHeight: 30,
    MenuBarHeight: 22,
    StatusBarHeight: 22,
    AnnotationMargin: 120,
    AnnotationRadius: 14,
    WindowPadding: 40,
    ChromelessPadding: 10,
} as const;

export interface ThemeColorValues {
    windowBackground: string;
    windowBorder: string;
    titleBarBackground: string;
    titleBarText: string;
    buttonBackground: string;
    buttonBorder: string;
    buttonText: string;
    textBoxBackground: string;
    textBoxBorder: string;
    textBoxText: string;
    textBoxPlaceholder: string;
    groupBorder: string;
    groupText: string;
    gridBackground: string;
    gridHeaderBackground: string;
    gridBorder: string;
    gridHeaderText: string;
    gridCellText: string;
    gridRowAlternate: string;
    menuBackground: string;
    menuText: string;
    menuBorder: string;
    statusBarBackground: string;
    statusBarText: string;
    labelText: string;
    linkLabelText: string;
    linkLabelUnderline: string;
    treeViewBackground: string;
    treeViewText: string;
    treeViewExpander: string;
    toolbarBackground: string;
    toolbarBorder: string;
    toolbarButtonHover: string;
    toolbarText: string;
    toolbarSeparator: string;
    annotationCircle: string;
    annotationText: string;
    annotationLine: string;
    chromelessBorder: string;
    canvasBackground: string;
}

const defaultColors: ThemeColorValues = {
    windowBackground: "#F0F0F0",
    windowBorder: "#999999",
    titleBarBackground: "#0078D4",
    titleBarText: "#FFFFFF",
    buttonBackground: "#E1E1E1",
    buttonBorder: "#ADADAD",
    buttonText: "#1E1E1E",
    textBoxBackground: "#FFFFFF",
    textBoxBorder: "#7A7A7A",
    textBoxText: "#1E1E1E",
    textBoxPlaceholder: "#A0A0A0",
    groupBorder: "#D5DFE5",
    groupText: "#333333",
    gridBackground: "#FFFFFF",
    gridHeaderBackground: "#F5F5F5",
    gridBorder: "#D5DFE5",
    gridHeaderText: "#1E1E1E",
    gridCellText: "#333333",
    gridRowAlternate: "#FAFAFA",
    menuBackground: "#F5F5F5",
    menuText: "#1E1E1E",
    menuBorder: "#D5DFE5",
    statusBarBackground: "#007ACC",
    statusBarText: "#FFFFFF",
    labelText: "#1E1E1E",
    linkLabelText: "#0066CC",
    linkLabelUnderline: "#0066CC",
    treeViewBackground: "#FFFFFF",
    treeViewText: "#1E1E1E",
    treeViewExpander: "#666666",
    toolbarBackground: "#F0F0F0",
    toolbarBorder: "#D5DFE5",
    toolbarButtonHover: "#E5E5E5",
    toolbarText: "#1E1E1E",
    toolbarSeparator: "#C0C0C0",
    annotationCircle: "#E53935",
    annotationText: "#FFFFFF",
    annotationLine: "#E53935",
    chromelessBorder: "#D0D0D0",
    canvasBackground: "#FFFFFF",
};

const darkColors: ThemeColorValues = {
    canvasBackground: "#1E1E1E",
    windowBackground: "#2D2D2D",
    windowBorder: "#555555",
    titleBarBackground: "#3C3C3C",
    titleBarText: "#CCCCCC",
    buttonBackground: "#3C3C3C",
    buttonBorder: "#555555",
    buttonText: "#CCCCCC",
    textBoxBackground: "#3C3C3C",
    textBoxBorder: "#555555",
    textBoxText: "#CCCCCC",
    textBoxPlaceholder: "#808080",
    groupBorder: "#555555",
    groupText: "#CCCCCC",
    gridBackground: "#2D2D2D",
    gridHeaderBackground: "#383838",
    gridBorder: "#555555",
    gridHeaderText: "#CCCCCC",
    gridCellText: "#BBBBBB",
    gridRowAlternate: "#333333",
    menuBackground: "#2D2D2D",
    menuText: "#CCCCCC",
    menuBorder: "#555555",
    statusBarBackground: "#007ACC",
    statusBarText: "#FFFFFF",
    labelText: "#CCCCCC",
    linkLabelText: "#4FC1FF",
    linkLabelUnderline: "#4FC1FF",
    treeViewBackground: "#2D2D2D",
    treeViewText: "#CCCCCC",
    treeViewExpander: "#AAAAAA",
    toolbarBackground: "#2D2D2D",
    toolbarBorder: "#555555",
    toolbarButtonHover: "#454545",
    toolbarText: "#CCCCCC",
    toolbarSeparator: "#555555",
    annotationCircle: "#E53935",
    annotationText: "#FFFFFF",
    annotationLine: "#E53935",
    chromelessBorder: "#555555",
};

const blueprintColors: ThemeColorValues = {
    canvasBackground: "#1A3A5C",
    windowBackground: "#1A3A5C",
    windowBorder: "#4A90D9",
    titleBarBackground: "#2E6EB5",
    titleBarText: "#FFFFFF",
    buttonBackground: "#1A3A5C",
    buttonBorder: "#4A90D9",
    buttonText: "#FFFFFF",
    textBoxBackground: "#0F2A45",
    textBoxBorder: "#4A90D9",
    textBoxText: "#FFFFFF",
    textBoxPlaceholder: "#7BAAD4",
    groupBorder: "#4A90D9",
    groupText: "#C0DEFF",
    gridBackground: "#0F2A45",
    gridHeaderBackground: "#1A3A5C",
    gridBorder: "#4A90D9",
    gridHeaderText: "#FFFFFF",
    gridCellText: "#C0DEFF",
    gridRowAlternate: "#15304F",
    menuBackground: "#1A3A5C",
    menuText: "#FFFFFF",
    menuBorder: "#4A90D9",
    statusBarBackground: "#2E6EB5",
    statusBarText: "#FFFFFF",
    labelText: "#FFFFFF",
    linkLabelText: "#7EC8E3",
    linkLabelUnderline: "#7EC8E3",
    treeViewBackground: "#0F2A45",
    treeViewText: "#FFFFFF",
    treeViewExpander: "#7BAAD4",
    toolbarBackground: "#1A3A5C",
    toolbarBorder: "#4A90D9",
    toolbarButtonHover: "#2E6EB5",
    toolbarText: "#FFFFFF",
    toolbarSeparator: "#4A90D9",
    annotationCircle: "#FF6B35",
    annotationText: "#FFFFFF",
    annotationLine: "#FF6B35",
    chromelessBorder: "#4A90D9",
};

const HEX_COLOR_RE = /^#[0-9A-Fa-f]{6}$/;

function createCustom(overrides: Record<string, string> | undefined): ThemeColorValues {
    const colors: ThemeColorValues = { ...defaultColors };
    if (!overrides) return colors;

    // Build case-insensitive property map
    const keyMap = new Map<string, keyof ThemeColorValues>();
    for (const k of Object.keys(colors) as (keyof ThemeColorValues)[]) {
        keyMap.set(k.toLowerCase(), k);
    }

    for (const [key, value] of Object.entries(overrides)) {
        const prop = keyMap.get(key.toLowerCase());
        if (prop && HEX_COLOR_RE.test(value)) {
            (colors as unknown as Record<string, string>)[prop] = value;
        }
    }

    return colors;
}

export class ThemeColors {
    private readonly _c: ThemeColorValues;

    constructor(colors?: ThemeColorValues) {
        this._c = colors ?? { ...defaultColors };
    }

    // ── accessors (match C# property names used in SvgRenderer) ──
    get WindowBackground() { return this._c.windowBackground; }
    get WindowBorder() { return this._c.windowBorder; }
    get TitleBarBackground() { return this._c.titleBarBackground; }
    get TitleBarText() { return this._c.titleBarText; }
    get ButtonBackground() { return this._c.buttonBackground; }
    get ButtonBorder() { return this._c.buttonBorder; }
    get ButtonText() { return this._c.buttonText; }
    get TextBoxBackground() { return this._c.textBoxBackground; }
    get TextBoxBorder() { return this._c.textBoxBorder; }
    get TextBoxText() { return this._c.textBoxText; }
    get TextBoxPlaceholder() { return this._c.textBoxPlaceholder; }
    get GroupBorder() { return this._c.groupBorder; }
    get GroupText() { return this._c.groupText; }
    get GridBackground() { return this._c.gridBackground; }
    get GridHeaderBackground() { return this._c.gridHeaderBackground; }
    get GridBorder() { return this._c.gridBorder; }
    get GridHeaderText() { return this._c.gridHeaderText; }
    get GridCellText() { return this._c.gridCellText; }
    get GridRowAlternate() { return this._c.gridRowAlternate; }
    get MenuBackground() { return this._c.menuBackground; }
    get MenuText() { return this._c.menuText; }
    get MenuBorder() { return this._c.menuBorder; }
    get StatusBarBackground() { return this._c.statusBarBackground; }
    get StatusBarText() { return this._c.statusBarText; }
    get LabelText() { return this._c.labelText; }
    get LinkLabelText() { return this._c.linkLabelText; }
    get LinkLabelUnderline() { return this._c.linkLabelUnderline; }
    get TreeViewBackground() { return this._c.treeViewBackground; }
    get TreeViewText() { return this._c.treeViewText; }
    get TreeViewExpander() { return this._c.treeViewExpander; }
    get ToolbarBackground() { return this._c.toolbarBackground; }
    get ToolbarBorder() { return this._c.toolbarBorder; }
    get ToolbarButtonHover() { return this._c.toolbarButtonHover; }
    get ToolbarText() { return this._c.toolbarText; }
    get ToolbarSeparator() { return this._c.toolbarSeparator; }
    get AnnotationCircle() { return this._c.annotationCircle; }
    get AnnotationText() { return this._c.annotationText; }
    get AnnotationLine() { return this._c.annotationLine; }
    get ChromelessBorder() { return this._c.chromelessBorder; }
    get CanvasBackground() { return this._c.canvasBackground; }

    static fromName(
        themeName: string | undefined,
        customOverrides?: Record<string, string>,
    ): ThemeColors {
        switch (themeName?.toLowerCase()) {
            case "dark":
                return new ThemeColors(darkColors);
            case "blueprint":
                return new ThemeColors(blueprintColors);
            case "custom":
                return new ThemeColors(createCustom(customOverrides));
            default:
                return new ThemeColors();
        }
    }
}

File: VehicleVision.Tools.ScreenSketch/Rendering/ThemeColors.cs
using System.Reflection;
using System.Text.RegularExpressions;

namespace VehicleVision.Tools.ScreenSketch.Rendering;

/// <summary>SVG 描画で使用する色定義。テーマごとにインスタンスを生成する</summary>
public partial class ThemeColors
{
    // ── Window ──
    public string WindowBackground { get; init; } = "#F0F0F0";
    public string WindowBorder { get; init; } = "#999999";
    public string TitleBarBackground { get; init; } = "#0078D4";
    public string TitleBarText { get; init; } = "#FFFFFF";

    // ── Button ──
    public string ButtonBackground { get; init; } = "#E1E1E1";
    public string ButtonBorder { get; init; } = "#ADADAD";
    public string ButtonText { get; init; } = "#1E1E1E";

    // ── TextBox ──
    public string TextBoxBackground { get; init; } = "#FFFFFF";
    public string TextBoxBorder { get; init; } = "#7A7A7A";
    public string TextBoxText { get; init; } = "#1E1E1E";
    public string TextBoxPlaceholder { get; init; } = "#A0A0A0";

    // ── GroupBox ──
    public string GroupBorder { get; init; } = "#D5DFE5";
    public string GroupText { get; init; } = "#333333";

    // ── DataGrid ──
    public string GridBackground { get; init; } = "#FFFFFF";
    public string GridHeaderBackground { get; init; } = "#F5F5F5";
    public string GridBorder { get; init; } = "#D5DFE5";
    public string GridHeaderText { get; init; } = "#1E1E1E";
    public string GridCellText { get; init; } = "#333333";
    public string GridRowAlternate { get; init; } = "#FAFAFA";

    // ── MenuBar ──
    public string MenuBackground { get; init; } = "#F5F5F5";
    public string MenuText { get; init; } = "#1E1E1E";
    public string MenuBorder { get; init; } = "#D5DFE5";

    // ── StatusBar ──
    public string StatusBarBackground { get; init; } = "#007ACC";
    public string StatusBarText { get; init; } = "#FFFFFF";

    // ── Label ──
    public string LabelText { get; init; } = "#1E1E1E";

    // ── LinkLabel ──
    public string LinkLabelText { get; init; } = "#0066CC";
    public string LinkLabelUnderline { get; init; } = "#0066CC";

    // ── TreeView ──
    public string TreeViewBackground { get; init; } = "#FFFFFF";
    public string TreeViewText { get; init; } = "#1E1E1E";
    public string TreeViewExpander { get; init; } = "#666666";

    // ── Toolbar ──
    public string ToolbarBackground { get; init; } = "#F0F0F0";
    public string ToolbarBorder { get; init; } = "#D5DFE5";
    public string ToolbarButtonHover { get; init; } = "#E5E5E5";
    public string ToolbarText { get; init; } = "#1E1E1E";
    public string ToolbarSeparator { get; init; } = "#C0C0C0";

    // ── Annotation ──
    public string AnnotationCircle { get; init; } = "#E53935";
    public string AnnotationText { get; init; } = "#FFFFFF";
    public string AnnotationLine { get; init; } = "#E53935";

    // ── Connector ──
    public string ConnectorLine { get; init; } = "#1976D2";
    public string ConnectorCircle { get; init; } = "#1976D2";
    public string ConnectorText { get; init; } = "#FFFFFF";

    // ── Chromeless ──
    public string ChromelessBorder { get; init; } = "#D0D0D0";

    // ── Canvas Background ──
    public string CanvasBackground { get; init; } = "#FFFFFF";

    /// <summary>テーマ名からプリセットを取得する</summary>
    /// <param name="themeName">テーマ名（default, dark, blueprint, custom）</param>
    /// <param name="customOverrides">カスタムテーマの色定義。theme が "custom" の場合に使用</param>
    public static ThemeColors FromName(string? themeName, Dictionary<string, string>? customOverrides = null)
    {
        return (themeName?.ToLowerInvariant()) switch
        {
            "dark" => CreateDark(),
            "blueprint" => CreateBlueprint(),
            "custom" => CreateCustom(customOverrides),
            _ => new ThemeColors()
        };
    }

    /// <summary>カスタムテーマ。未指定の要素は標準テーマにフォールバックする</summary>
    public static ThemeColors CreateCustom(Dictionary<string, string>? overrides)
    {
        var colors = new ThemeColors();
        if (overrides is not { Count: > 0 }) return colors;

        foreach (var (key, value) in overrides)
        {
            if (ColorPropertyMap.TryGetValue(key, out var prop) && HexColorRegex().IsMatch(value))
            {
                prop.SetValue(colors, value);
            }
        }

        return colors;
    }

    /// <summary>色プロパティのキャッシュ（大文字小文字を区別しないキー）</summary>
    private static readonly Dictionary<string, PropertyInfo> ColorPropertyMap =
        typeof(ThemeColors).GetProperties(BindingFlags.Public | BindingFlags.Instance)
            .Where(p => p.PropertyType == typeof(string) && p.CanWrite)
            .ToDictionary(p => p.Name, p => p, StringComparer.OrdinalIgnoreCase);

    [GeneratedRegex(@"^#[0-9A-Fa-f]{6}$")]
    private static partial Regex HexColorRegex();

    /// <summary>ダークテーマ</summary>
    public static ThemeColors CreateDark() => new()
    {
        CanvasBackground = "#1E1E1E",
        WindowBackground = "#2D2D2D",
        WindowBorder = "#555555",
        TitleBarBackground = "#3C3C3C",
        TitleBarText = "#CCCCCC",
        ButtonBackground = "#3C3C3C",
        ButtonBorder = "#555555",
        ButtonText = "#CCCCCC",
        TextBoxBackground = "#3C3C3C",
        TextBoxBorder = "#555555",
        TextBoxText = "#CCCCCC",
        TextBoxPlaceholder = "#808080",
        GroupBorder = "#555555",
        GroupText = "#CCCCCC",
        GridBackground = "#2D2D2D",
        GridHeaderBackground = "#383838",
        GridBorder = "#555555",
        GridHeaderText = "#CCCCCC",
        GridCellText = "#BBBBBB",
        GridRowAlternate = "#333333",
        MenuBackground = "#2D2D2D",
        MenuText = "#CCCCCC",
        MenuBorder = "#555555",
        StatusBarBackground = "#007ACC",
        StatusBarText = "#FFFFFF",
        LabelText = "#CCCCCC",
        LinkLabelText = "#4FC1FF",
        LinkLabelUnderline = "#4FC1FF",
        TreeViewBackground = "#2D2D2D",
        TreeViewText = "#CCCCCC",
        TreeViewExpander = "#AAAAAA",
        ToolbarBackground = "#2D2D2D",
        ToolbarBorder = "#555555",
        ToolbarButtonHover = "#454545",
        ToolbarText = "#CCCCCC",
        ToolbarSeparator = "#555555",
        AnnotationCircle = "#E53935",
        AnnotationText = "#FFFFFF",
        AnnotationLine = "#E53935",
        ConnectorLine = "#42A5F5",
        ConnectorCircle = "#42A5F5",
        ConnectorText = "#FFFFFF",
        ChromelessBorder = "#555555"
    };

    /// <summary>設計図風ブループリントテーマ</summary>
    public static ThemeColors CreateBlueprint() => new()
    {
        CanvasBackground = "#1A3A5C",
        WindowBackground = "#1A3A5C",
        WindowBorder = "#4A90D9",
        TitleBarBackground = "#2E6EB5",
        TitleBarText = "#FFFFFF",
        ButtonBackground = "#1A3A5C",
        ButtonBorder = "#4A90D9",
        ButtonText = "#FFFFFF",
        TextBoxBackground = "#0F2A45",
        TextBoxBorder = "#4A90D9",
        TextBoxText = "#FFFFFF",
        TextBoxPlaceholder = "#7BAAD4",
        GroupBorder = "#4A90D9",
        GroupText = "#C0DEFF",
        GridBackground = "#0F2A45",
        GridHeaderBackground = "#1A3A5C",
        GridBorder = "#4A90D9",
        GridHeaderText = "#FFFFFF",
        GridCellText = "#C0DEFF",
        GridRowAlternate = "#15304F",
        MenuBackground = "#1A3A5C",
        MenuText = "#FFFFFF",
        MenuBorder = "#4A90D9",
        StatusBarBackground = "#2E6EB5",
        StatusBarText = "#FFFFFF",
        LabelText = "#FFFFFF",
        LinkLabelText = "#7EC8E3",
        LinkLabelUnderline = "#7EC8E3",
        TreeViewBackground = "#0F2A45",
        TreeViewText = "#FFFFFF",
        TreeViewExpander = "#7BAAD4",
        ToolbarBackground = "#1A3A5C",
        ToolbarBorder = "#4A90D9",
        ToolbarButtonHover = "#2E6EB5",
        ToolbarText = "#FFFFFF",
        ToolbarSeparator = "#4A90D9",
        AnnotationCircle = "#FF6B35",
        AnnotationText = "#FFFFFF",
        AnnotationLine = "#FF6B35",
        ConnectorLine = "#7EC8E3",
        ConnectorCircle = "#7EC8E3",
        ConnectorText = "#FFFFFF",
        ChromelessBorder = "#4A90D9"
    };
}
